CHANGELOG.md

```diff
@@ -40,6 +41,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - When `PretrainedTransformerIndexer` folds long sequences, it no longer loses the information from token type ids.
 - Fixed documentation for `GradientDescentTrainer.cuda_device`.
+- Re-starting a training run from a checkpoint in the middle of an epoch now works correctly.
+- When using the "moving average" weights smoothing feature of the trainer, training checkpoints would also get smoothed, with strange results for resuming a training job. This has been fixed.
+- When re-starting an interrupted training job, the trainer will now read out the data loader even for epochs and batches that can be skipped. This ensures that any random number generators used by the reader or data loader are in the same state as they were the first time the training job ran.
```

Comment:
This sounds good, in theory, but there are probably other things that affect the random number generators used by the reader and data loader. I don't think we can guarantee the same order.

Reply:
Hmm. I wrote it this way because in Quark it worked out that way. I had good enough control over the RNGs that it was deterministic.

In AllenNLP, we can't guarantee that none of the things we're skipping when restoring from a checkpoint (the `forward()` method for example) modify the RNG state. I guess I'll say that this is an attempt to ensure deterministic randomness, but does not guarantee it. At the same time, we should encourage components to use their own RNG instead of using the global one, so they don't affect each other.

Reply:
It's actually quite bad if this doesn't work. If we don't guarantee the order of instances, and we stop training 10 times in the middle of an epoch and restart it, we might end up training on the same instance 10 times.

Reply:
11 times even